### PR TITLE
Switch map_index back to use server_default

### LINKED ISSUE #20876: Airflow database upgrade fails with "psycopg2.errors.NotNullViolation: column "map_index" of relation "task_instance" contains null value"s
### Apache Airflow version

main (development)

### What happened

I currently have Airflow 2.2.3 and due to this [issue](https://github.com/apache/airflow/issues/19699) I have tried to upgrade Airflow to this [commit](https://github.com/apache/airflow/commit/14ee831c7ad767e31a3aeccf3edbc519b3b8c923).

When I run `airflow db upgrade` I get the following error:
```
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 587bdf053233 -> e655c0453f75, Add TaskMap and map_index on TaskInstance.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/sqlalchemy/engine/base.py", line 1277, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.7/dist-packages/sqlalchemy/engine/default.py", line 608, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.NotNullViolation: column "map_index" of relation "task_instance" contains null values
```

The map_index column was introduced with this [PR](https://github.com/apache/airflow/pull/20286).

Could you please advise? 

### What you expected to happen

_No response_

### How to reproduce

_No response_

### Operating System

Ubuntu 18.04.6 LTS

### Versions of Apache Airflow Providers

_No response_

### Deployment

Other

### Deployment details

Kubernetes

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by uranusjr**
Close #20876.

This ends up being simply reverting fff3a1e866bfc2d00d6454a5feb72203a8ce7513, but I forgot exactly why I made the switch in the first place. I guess the test suite will tell us if it works…

### CODE REVIEW

**Comment by ashb** [comment]
Not _quite_ related to this, but when trying the migration from 2.2.3 to latest main I also get:

```sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedObject) constraint "task_reschedule_ti_fkey" of relation "task_reschedule" does not exist
```

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -343,7 +344,7 @@ class TaskInstance(Base, LoggingMixin):
     task_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True, nullable=False)
     dag_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True, nullable=False)
     run_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True, nullable=False)
-    map_index = Column(Integer, primary_key=True, nullable=False, default=-1)
+    map_index = Column(Integer, primary_key=True, nullable=False, server_default=text("-1"))
```
Comment (ashb) [inline airflow/models/taskinstance.py:347]: This one can possibly have default and server_default.

**Review by ashb — APPROVED** [review]
Migration change works.

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (uranusjr) [reply airflow/models/taskinstance.py:347]: I checked some sources and realised `default` isn’t actually much useful in ORM, only if we use the SQL expression builder. And with `server_default` set, `default` only really matters for `UPDATE`, which actually seems a bit wrong to me, so I’ll omit `default` until they are proven useful.

**Pull request merged**